File: inspirations/extensions/flow-webext/src/runtime/catalog.ts
import type { BrowserPackManifest } from "./protocol";

export const FLOW_BROWSERPACK_ORIGIN = "https://flow.browserpack.local";

export const browserPackCatalog: BrowserPackManifest[] = [
  {
    version: 1,
    packKey: "qwen3-0.6b-onnx-browserpack",
    modelKey: "qwen3-0.6b",
    displayName: "Qwen3 0.6B Browser Pack",
    repoId: "onnx-community/Qwen3-0.6B-DQ-ONNX",
    modality: "chat",
    backend: "transformersjs-onnx",
    quantization: "q4f16",
    requiresWebgpu: false,
    files: [
      {
        path: "config.json",
        sourceUrl:
          "https://huggingface.co/onnx-community/Qwen3-0.6B-DQ-ONNX/resolve/main/config.json",
        purpose: "model-config",
        contentType: "application/json",
      },
      {
        path: "generation_config.json",
        sourceUrl:
          "https://huggingface.co/onnx-community/Qwen3-0.6B-DQ-ONNX/resolve/main/generation_config.json",
        purpose: "generation-config",
        contentType: "application/json",
      },
      {
        path: "tokenizer.json",
        sourceUrl:
          "https://huggingface.co/onnx-community/Qwen3-0.6B-DQ-ONNX/resolve/main/tokenizer.json",
        purpose: "tokenizer",
        contentType: "application/json",
      },
      {
        path: "tokenizer_config.json",
        sourceUrl:
          "https://huggingface.co/onnx-community/Qwen3-0.6B-DQ-ONNX/resolve/main/tokenizer_config.json",
        purpose: "tokenizer-config",
        contentType: "application/json",
      },
      {
        path: "special_tokens_map.json",
        sourceUrl:
          "https://huggingface.co/onnx-community/Qwen3-0.6B-DQ-ONNX/resolve/main/special_tokens_map.json",
        purpose: "special-tokens",
        contentType: "application/json",
      },
      {
        path: "onnx/model_q4f16.onnx",
        sourceUrl:
          "https://huggingface.co/onnx-community/Qwen3-0.6B-DQ-ONNX/resolve/main/onnx/model_q4f16.onnx",
        purpose: "decoder-model",
        contentType: "application/octet-stream",
      },
      {
        path: "onnx/model.onnx_data",
        sourceUrl:
          "https://huggingface.co/onnx-community/Qwen3-0.6B-DQ-ONNX/resolve/main/onnx/model.onnx_data",
        purpose: "decoder-weights",
        contentType: "application/octet-stream",
      },
    ],
  },
  {
    version: 1,
    packKey: "trocr-small-printed-browserpack",
    modelKey: "trocr-small-printed",
    displayName: "TrOCR Small Printed Browser Pack",
    repoId: "Xenova/trocr-small-printed",
    modality: "ocr",
    backend: "transformersjs-onnx",
    quantization: "quantized",
    requiresWebgpu: false,
    files: [
      {
        path: "config.json",
        sourceUrl: "https://huggingface.co/Xenova/trocr-small-printed/resolve/main/config.json",
        purpose: "model-config",
        contentType: "application/json",
      },
      {
        path: "preprocessor_config.json",
        sourceUrl:
          "https://huggingface.co/Xenova/trocr-small-printed/resolve/main/preprocessor_config.json",
        purpose: "processor-config",
        contentType: "application/json",
      },
      {
        path: "tokenizer_config.json",
        sourceUrl:
          "https://huggingface.co/Xenova/trocr-small-printed/resolve/main/tokenizer_config.json",
        purpose: "tokenizer-config",
        contentType: "application/json",
      },
      {
        path: "vocab.json",
        sourceUrl: "https://huggingface.co/Xenova/trocr-small-printed/resolve/main/vocab.json",
        purpose: "tokenizer-vocab",
        contentType: "application/json",
      },
      {
        path: "onnx/encoder_model_quantized.onnx",
        sourceUrl:
          "https://huggingface.co/Xenova/trocr-small-printed/resolve/main/onnx/encoder_model_quantized.onnx",
        purpose: "ocr-encoder",
        contentType: "application/octet-stream",
      },
      {
        path: "onnx/decoder_model_merged_quantized.onnx",
        sourceUrl:
          "https://huggingface.co/Xenova/trocr-small-printed/resolve/main/onnx/decoder_model_merged_quantized.onnx",
        purpose: "ocr-decoder",
        contentType: "application/octet-stream",
      },
    ],
  },
  {
    version: 1,
    packKey: "qwen3.5-0.8b-onnx-browserpack",
    modelKey: "qwen3.5-0.8b",
    displayName: "Qwen3.5 0.8B Browser Pack",
    repoId: "onnx-community/Qwen3.5-0.8B-ONNX",
    modality: "vision-language",
    backend: "transformersjs-onnx",
    quantization: "mixed-q4-fp16",
    requiresWebgpu: true,
    files: [
      {
        path: "config.json",
        sourceUrl:
          "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/config.json",
        purpose: "model-config",
        contentType: "application/json",
      },
      {
        path: "generation_config.json",
        sourceUrl:
          "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/generation_config.json",
        purpose: "generation-config",
        contentType: "application/json",
      },
      {
        path: "processor_config.json",
        sourceUrl:
          "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/processor_config.json",
        purpose: "processor-config",
        contentType: "application/json",
      },
      {
        path: "preprocessor_config.json",
        sourceUrl:
          "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/preprocessor_config.json",
        purpose: "preprocessor-config",
        contentType: "application/json",
      },
      {
        path: "tokenizer.json",
        sourceUrl:
          "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/tokenizer.json",
        purpose: "tokenizer",
        contentType: "application/json",
      },
      {
        path: "tokenizer_config.json",
        sourceUrl:
          "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/tokenizer_config.json",
        purpose: "tokenizer-config",
        contentType: "application/json",
      },
      {
        path: "onnx/embed_tokens_q4.onnx",
        sourceUrl:
          "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/onnx/embed_tokens_q4.onnx",
        purpose: "embed-tokens",
        contentType: "application/octet-stream",
      },
      {
        path: "onnx/vision_encoder_fp16.onnx",
        sourceUrl:
          "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/onnx/vision_encoder_fp16.onnx",
        purpose: "vision-encoder",
        contentType: "application/octet-stream",
      },
      {
        path: "onnx/decoder_model_merged_q4.onnx",
        sourceUrl:
          "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/onnx/decoder_model_merged_q4.onnx",
        purpose: "decoder-model",
        contentType: "application/octet-stream",
      },
    ],
  },
];

export function getBrowserPackByModelKey(modelKey: string) {
  return browserPackCatalog.find((pack) => pack.modelKey === modelKey) ?? null;
}

export function modelBaseUrl(packKey: string) {
  return `${FLOW_BROWSERPACK_ORIGIN}/${packKey}`;
}

File: inspirations/src/browser/catalog.rs
use crate::runtime::Modality;

use super::types::{
    BrowserExecutionBackend, BrowserPackFile, BrowserPackManifest, BrowserPackSupport,
};

fn pack(
    pack_key: &str,
    model_key: &str,
    display_name: &str,
    repo_id: &str,
    modality: Modality,
    backend: BrowserExecutionBackend,
    quantization: Option<&str>,
    tokenizer: Option<&str>,
    processor: Option<&str>,
    browser_support: BrowserPackSupport,
    files: &[(&str, &str, Option<u64>, &str)],
    tags: &[&str],
) -> BrowserPackManifest {
    BrowserPackManifest {
        version: 1,
        pack_key: pack_key.to_string(),
        model_key: model_key.to_string(),
        display_name: display_name.to_string(),
        repo_id: repo_id.to_string(),
        modality,
        backend,
        quantization: quantization.map(str::to_string),
        tokenizer: tokenizer.map(str::to_string),
        processor: processor.map(str::to_string),
        browser_support,
        files: files
            .iter()
            .map(|(path, source_url, bytes, purpose)| BrowserPackFile {
                path: (*path).to_string(),
                source_url: (*source_url).to_string(),
                bytes: *bytes,
                sha256: None,
                required: true,
                purpose: (*purpose).to_string(),
            })
            .collect(),
        tags: tags.iter().map(|tag| (*tag).to_string()).collect(),
    }
}

pub fn default_browser_pack_catalog() -> Vec<BrowserPackManifest> {
    vec![
        pack(
            "qwen3-0.6b-onnx-browserpack",
            "qwen3-0.6b",
            "Qwen3 0.6B Browser Pack",
            "onnx-community/Qwen3-0.6B-DQ-ONNX",
            Modality::Chat,
            BrowserExecutionBackend::TransformersJsOnnx,
            Some("q4f16"),
            Some("tokenizer.json"),
            None,
            BrowserPackSupport {
                chromium: true,
                firefox: true,
                safari: true,
                standalone_web: true,
                requires_webgpu: false,
            },
            &[
                (
                    "config.json",
                    "https://huggingface.co/onnx-community/Qwen3-0.6B-DQ-ONNX/resolve/main/config.json",
                    None,
                    "model-config",
                ),
                (
                    "generation_config.json",
                    "https://huggingface.co/onnx-community/Qwen3-0.6B-DQ-ONNX/resolve/main/generation_config.json",
                    None,
                    "generation-config",
                ),
                (
                    "tokenizer.json",
                    "https://huggingface.co/onnx-community/Qwen3-0.6B-DQ-ONNX/resolve/main/tokenizer.json",
                    None,
                    "tokenizer",
                ),
                (
                    "tokenizer_config.json",
                    "https://huggingface.co/onnx-community/Qwen3-0.6B-DQ-ONNX/resolve/main/tokenizer_config.json",
                    None,
                    "tokenizer-config",
                ),
                (
                    "special_tokens_map.json",
                    "https://huggingface.co/onnx-community/Qwen3-0.6B-DQ-ONNX/resolve/main/special_tokens_map.json",
                    None,
                    "special-tokens",
                ),
                (
                    "onnx/model_q4f16.onnx",
                    "https://huggingface.co/onnx-community/Qwen3-0.6B-DQ-ONNX/resolve/main/onnx/model_q4f16.onnx",
                    None,
                    "decoder-model",
                ),
                (
                    "onnx/model.onnx_data",
                    "https://huggingface.co/onnx-community/Qwen3-0.6B-DQ-ONNX/resolve/main/onnx/model.onnx_data",
                    None,
                    "decoder-weights",
                ),
            ],
            &["text", "cross-browser", "baseline", "local-first"],
        ),
        pack(
            "trocr-small-printed-browserpack",
            "trocr-small-printed",
            "TrOCR Small Printed Browser Pack",
            "Xenova/trocr-small-printed",
            Modality::Ocr,
            BrowserExecutionBackend::TransformersJsOnnx,
            Some("quantized"),
            None,
            Some("preprocessor_config.json"),
            BrowserPackSupport {
                chromium: true,
                firefox: true,
                safari: true,
                standalone_web: true,
                requires_webgpu: false,
            },
            &[
                (
                    "config.json",
                    "https://huggingface.co/Xenova/trocr-small-printed/resolve/main/config.json",
                    None,
                    "model-config",
                ),
                (
                    "preprocessor_config.json",
                    "https://huggingface.co/Xenova/trocr-small-printed/resolve/main/preprocessor_config.json",
                    None,
                    "processor-config",
                ),
                (
                    "tokenizer_config.json",
                    "https://huggingface.co/Xenova/trocr-small-printed/resolve/main/tokenizer_config.json",
                    None,
                    "tokenizer-config",
                ),
                (
                    "vocab.json",
                    "https://huggingface.co/Xenova/trocr-small-printed/resolve/main/vocab.json",
                    None,
                    "tokenizer-vocab",
                ),
                (
                    "onnx/encoder_model_quantized.onnx",
                    "https://huggingface.co/Xenova/trocr-small-printed/resolve/main/onnx/encoder_model_quantized.onnx",
                    None,
                    "ocr-encoder",
                ),
                (
                    "onnx/decoder_model_merged_quantized.onnx",
                    "https://huggingface.co/Xenova/trocr-small-printed/resolve/main/onnx/decoder_model_merged_quantized.onnx",
                    None,
                    "ocr-decoder",
                ),
            ],
            &["ocr", "cross-browser", "document", "image"],
        ),
        pack(
            "qwen3.5-0.8b-onnx-browserpack",
            "qwen3.5-0.8b",
            "Qwen3.5 0.8B Browser Pack",
            "onnx-community/Qwen3.5-0.8B-ONNX",
            Modality::VisionLanguage,
            BrowserExecutionBackend::TransformersJsOnnx,
            Some("q4"),
            Some("tokenizer.json"),
            Some("processor_config.json"),
            BrowserPackSupport {
                chromium: true,
                firefox: true,
                safari: true,
                standalone_web: true,
                requires_webgpu: true,
            },
            &[
                (
                    "config.json",
                    "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/config.json",
                    None,
                    "model-config",
                ),
                (
                    "generation_config.json",
                    "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/generation_config.json",
                    None,
                    "generation-config",
                ),
                (
                    "processor_config.json",
                    "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/processor_config.json",
                    None,
                    "processor-config",
                ),
                (
                    "preprocessor_config.json",
                    "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/preprocessor_config.json",
                    None,
                    "preprocessor-config",
                ),
                (
                    "tokenizer.json",
                    "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/tokenizer.json",
                    None,
                    "tokenizer",
                ),
                (
                    "tokenizer_config.json",
                    "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/tokenizer_config.json",
                    None,
                    "tokenizer-config",
                ),
                (
                    "onnx/embed_tokens_q4.onnx",
                    "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/onnx/embed_tokens_q4.onnx",
                    None,
                    "embed-tokens",
                ),
                (
                    "onnx/vision_encoder_fp16.onnx",
                    "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/onnx/vision_encoder_fp16.onnx",
                    None,
                    "vision-encoder",
                ),
                (
                    "onnx/decoder_model_merged_q4.onnx",
                    "https://huggingface.co/onnx-community/Qwen3.5-0.8B-ONNX/resolve/main/onnx/decoder_model_merged_q4.onnx",
                    None,
                    "decoder-model",
                ),
            ],
            &["multimodal", "image", "document", "webgpu"],
        ),
    ]
}
